### DISCUSSION TITLE
Explicit Typing of `getContract`

### DISCUSSION THREAD

**Discussion by hickscorp** [wevm/wagmi#2375] [Ideas]
Consider this:

```ts
const contract = useMemo(() => getContract({ abi, address }), [address]);
```

At this point, `contract` is typed and all its accessors are typed as well (And BTW it's really great, we're migrating a big app from `0.12` to `1.0` and it's such a pleasure).
The problem at this point seems to be that there's no easy way to formulate the type of `contract`, so if a function needs to receive it... It's fairly cumbersome.
It would be nice to see something like:

```ts
const contract: Contract<typeof abi> = useMemo(() => getContract({ abi, address }), [address]);
```

We tried to play a little bit with `ReturnType<typeof getContract>` but we got nowhere.

Any idea, or should we suggest that this could become a feature?

**Comment by tmm** [comment]
Can accomplish this in a couple ways.

#### viem `GetContractReturnType`

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzjAnmApnA4umBhCAOxigEMBjGAJVwFcpCAVNdAGmRbgAVaAjAG2Dk8g9MXaoMcAOql+-XCOBj4AXzgAzKBBBwA5ADdlIPQChQkWIjjoo5AEwAGAIK9gcdVp37SbyegB6GHQAZxgzU384AFkUAmIySjgAXmxFIhIKajoGZgwAHn8IDRs7J1dgdh4BISUVdll5RVFiAD4gA)

```ts
import { type GetContractReturnType, type PublicClient, type WalletClient } from 'viem'
import { erc20Abi } from 'abitype/test'

type MyContract = GetContractReturnType<typeof erc20Abi, PublicClient, WalletClient>
```

#### wagmi `GetContractResult`

Import `GetContractResult` from `@wagmi/core` or `wagmi/actions (wrapper around viem's `GetContractReturnType`)

[TypeScript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzjAnmApnA6gQwDZ7owDCew6AdvAL5wBmUEIcA5AG7kgsBQoksiZGkwBxIsQhUoOAMYwASugDOAVzw16jZiwACAdxwBzEMAD0M6Oh59o8JOigyATAAYAggCNgcWgyascL1QMUxhlGB5uYMwAWRQJKVl4AF44MRJJGGk5RVV1AB5oiDo4B2d3LwAabHxCEjJKGAA+IA)

```ts
import { type WalletClient } from 'viem'
import { type GetContractResult } from '@wagmi/core'
import { erc20Abi } from 'abitype/test'

type MyContract = GetContractResult<typeof erc20Abi, WalletClient>
```
